File: TrackingVisualizer/TTRFile.py
from typing import Union
import io

from Logger import Logger

class TTRFile:
    domains: list[str]
    entities: list[tuple[int, str]]
    
    def __init__(self, path: str):
        self.__path = path
        self.__file = None
        self.__has_parsed = False
        
        if not self.__open():
            Logger.append_info("Failed to open file: {}", path)
            return
            
        if not self.__parse():
            Logger.append_info("Failed to parse file: {}", path)
            return
            
    def __del__(self):
        if not self.__close():
            Logger.append_info("Failed to close file: {}", self.__path)
            
    def ready(self) -> bool:
        return self.__has_parsed
    
    ## Domain handling
    
    def domain_exists(self, domain: str) -> bool:
        if not self.ready():
            Logger.log_error("Invalid state")
            return False
            
        return domain in self.domains
    
    def domain_id_exists(self, domain_id: int) -> bool:
        if not self.ready():
            Logger.log_error("Invalid state")
            return False
            
        return domain_id < len(self.domains)
    
    def get_domain_id(self, domain: str) -> Union[int, None]:
        if not self.ready():
            Logger.log_error("Invalid state")
            return None
            
        if not self.domain_exists(domain):
            Logger.log_error("Domain does not exist: {}", domain)
            return None
            
        return self.domains.index(domain)
    
    def get_domain(self, domain_id: int) -> Union[str, None]:
        if not self.ready():
            Logger.log_error("Invalid state")
            return None
            
        if not self.domain_id_exists(domain_id):
            Logger.log_error("Domain ID does not exist: {}", domain_id)
            return None
            
        return self.domains[domain_id]
    
    
    ## Entity handling
    
    def entity_exists(self, entity: str, domain_id: int) -> bool:
        if not self.ready():
            Logger.log_error("Invalid state")
            return False
            
        return (domain_id, entity) in self.entities
    
    def entity_id_exists(self, entity_id: int) -> bool:
        if not self.ready():
            Logger.log_error("Invalid state")
            return False
            
        return entity_id < len(self.entities)     
    
    def get_entity_id(self, entity: str, domain_id: int) -> Union[int, None]:
        if not self.ready():
            Logger.log_error("Invalid state")
            return None
            
        if not self.entity_exists(entity, domain_id):
            Logger.log_error("Entity does not exist: {}", entity)
            return None
            
        return self.entities.index((domain_id, entity))
    
    def get_entity(self, entity_id: int) -> Union[tuple[int, str], None]:
        if not self.ready():
            Logger.log_error("Invalid state")
            return None
            
        if not self.entity_id_exists(entity_id):
            Logger.log_error("Entity ID does not exist: {}", entity_id)
            return None
            
        return self.entities[entity_id]
    
    
    ## File handling
    
    __path: str
    __file: io.BufferedReader
    
    __has_parsed: bool
    
    def __open(self) -> bool:
        try:
            self.__file = open(self.__path, "rb")
            return True
        except FileNotFoundError:
            Logger.log_error("File not found")
            return False
        except PermissionError:
            Logger.log_error("Permission denied")
            return False
        except:
            Logger.log_error("Unknown error")
            return False
    
    def __close(self) -> bool:
        if self.__file is None:
            Logger.log_warning("File already closed")
            return True
            
        try:
            self.__file.close()
            return True
        except:
            Logger.log_error("Unknown error")
            return False
    
    def __parse(self) -> bool:
        if not self.__file and not self.__open():
            Logger.append_info("Failed to parse file: {}", self.__path)
            return False
            
        self.domains = []
        self.entities = []
        
        try:
            self.__file.seek(0)
            header: str = self.__file.read(3).decode("utf-8")
            
            if header != "TTR":
                Logger.log_error("Invalid file header: {}", header)
                return False
                
            offset_to_domains_start: int = int.from_bytes(self.__file.read(4), byteorder="little")
            offset_to_domains_end: int = int.from_bytes(self.__file.read(4), byteorder="little")
            
            offset_to_entities: int = int.from_bytes(self.__file.read(4), byteorder="little")
            
            self.__file.seek(offset_to_domains_start)
            
            num_of_domains: int = int.from_bytes(self.__file.read(1), byteorder="little")
            
            for i in range(num_of_domains):
                length: int = int.from_bytes(self.__file.read(1), byteorder="little")
                domain: str = self.__file.read(length).decode("utf-8")
                self.domains.append(domain)
                
            self.__file.seek(offset_to_entities)
            
            num_of_entities: int = int.from_bytes(self.__file.read(2), byteorder="little")
            
            for i in range(num_of_entities):
                domain: int = int.from_bytes(self.__file.read(1), byteorder="little")
                length: int = int.from_bytes(self.__file.read(1), byteorder="little")
                entity: str = self.__file.read(length).decode("utf-8")
                self.entities.append((domain, entity))
                
            self.__has_parsed = True
            return True
            
        except Exception as e:
            Logger.log_error("An error occurred while parsing: {}", e)
            return False
        

File: TimeTracker/src/Database/TTRFile.cpp
#include "TTRFile.h"

TTRFile::TTRFile(const std::wstring& file_path)
	: _file_path(file_path), _file()
{
}

TTRFile::~TTRFile()
{
	if (_file.is_open())
	{
		_close();
	}
}

bool TTRFile::add_domain(const std::string& domain)
{
	if (!_file.is_open() && !_open())
	{
		return false;
	}

	if (!_read_info())
	{
		return false;
	}

	if (domain_exists(domain))
	{
		Logger::log_warning("Domain already exists: {}", domain);
		return false;
	}

	size_t required_size = domain.size() + 1;

	if (_offset_to_entities - _offset_to_domains_end < required_size)
	{
		// TODO: Implement block reallocation
		Logger::log_error("Not enough space to add domain: {}", domain);
		return false;
	}
	else
	{
		_file.seekp(_offset_to_domains_end);
		
		uint8_t len = domain.size();
		_file.write(reinterpret_cast<const char*>(&len), sizeof(len));

		_file.write(domain.c_str(), len);

		_offset_to_domains_end += required_size;

		_num_of_domains++;
		_file.seekg(_offset_to_current_domain);
		_file.write(reinterpret_cast<const char*>(&_num_of_domains), sizeof(_num_of_domains));

		_update_header();
	}

	_close();

	return true;
}

TTRFile::domain_id TTRFile::get_domain_id(const std::string& domain)
{
	if (!_file.is_open() && !_open())
	{
		return false;
	}

	if (!_read_info())
	{
		return false;
	}

	_file.seekg(_offset_to_current_domain);

	uint8_t num_of_domains;
	_file.read(reinterpret_cast<char*>(&num_of_domains), sizeof(num_of_domains));


	for (uint8_t i = 0; i < num_of_domains; i++)
	{
		uint8_t len;
		_file.read(reinterpret_cast<char*>(&len), sizeof(len));

		std::string name(len, '\0');
		_file.read(name.data(), len);

		if (name == domain)
		{
			return i;
		}
	}

	return -1;
}

bool TTRFile::domain_exists(const std::string& domain)
{
	return get_domain_id(domain) != (domain_id)-1;
}

bool TTRFile::add_entity(domain_id id, const std::string& entity)
{
	if (!_file.is_open() && !_open())
	{
		return false;
	}

	if (!_read_info())
	{
		return false;
	}

	if (entity_exists(id, entity))
	{
		Logger::log_warning("Entity already exists: {}", entity);
		return false;
	}

	_file.seekg(_offset_to_entities);

	uint16_t num_of_entities;
	_file.read(reinterpret_cast<char*>(&num_of_entities), sizeof(num_of_entities));

	num_of_entities++;

	_file.seekp(_offset_to_entities);
	_file.write(reinterpret_cast<const char*>(&num_of_entities), sizeof(num_of_entities));

	_file.seekp(0, std::ios::end);

	_file.write(reinterpret_cast<const char*>(&id), sizeof(id));

	uint8_t len = entity.size();
	_file.write(reinterpret_cast<const char*>(&len), sizeof(len));

	_file.write(entity.c_str(), len);

	return true;
}

TTRFile::entity_id TTRFile::get_entity_id(domain_id id, const std::string& entity)
{
	if (!_file.is_open() && !_open())
	{
		return false;
	}

	if (!_read_info())
	{
		return false;
	}

	_file.seekg(_offset_to_entities);

	uint16_t num_of_entities;
	_file.read(reinterpret_cast<char*>(&num_of_entities), sizeof(num_of_entities));

	for (uint16_t i = 0; i < num_of_entities; i++)
	{
		domain_id domain_id;
		_file.read(reinterpret_cast<char*>(&domain_id), sizeof(domain_id));

		uint8_t len;
		_file.read(reinterpret_cast<char*>(&len), sizeof(len));

		std::string name(len, '\0');
		_file.read(name.data(), len);

		if (domain_id == id && name == entity)
		{
			return i;
		}
	}

	return -1;
}

bool TTRFile::entity_exists(domain_id id, const std::string& entity)
{
	return get_entity_id(id, entity) != (entity_id)-1;
}

bool TTRFile::_open()
{
	if (!std::filesystem::exists(_file_path))
	{
		if (!_create())
		{
			return false;
		}
	}

	_file.open(_file_path, std::ios::in | std::ios::out | std::ios::binary);

	if (!_file.is_open())
	{
		Logger::log_error("Failed to open file: {}", (char*)_file_path.c_str());
		return false;
	}

	return true;
}

bool TTRFile::_close()
{
	_file.close();

	if (_file.is_open())
	{
		Logger::log_error("Failed to close file: {}", (char*)_file_path.c_str());
		return false;
	}

	return true;
}

bool TTRFile::_create()
{
	if (!std::filesystem::exists(_file_path))
	{
		std::filesystem::create_directories(std::filesystem::path(_file_path).parent_path());
	}

	_file.open(_file_path, std::ios::out | std::ios::binary);

	if (!_file.is_open())
	{
		char error_buffer[256];
		strerror_s(error_buffer, 256, errno);

		Logger::log_error("Failed to create file: {}", error_buffer);
		return false;
	}

	_offset_to_current_domain = 15;
	_offset_to_domains_end = 16;
	_num_of_domains = 0;

	_offset_to_entities = _offset_to_domains_end + DOMAINS_MIN_BLOCK_SIZE;
	_num_of_entities = 0;

	_update_header();

	_file.seekg(_offset_to_current_domain);
	_file.write(reinterpret_cast<const char*>(&_num_of_domains), sizeof(_num_of_domains));

	for (int i = 0; i < DOMAINS_MIN_BLOCK_SIZE; i++)
	{
		_file.write("#", 1);
	}

	_file.seekg(_offset_to_entities);

	_file.write(reinterpret_cast<const char*>(&_num_of_entities), sizeof(_num_of_entities));

	_close();

	return true;
}

bool TTRFile::_read_info()
{
	_file.seekg(0);

	char header[4];
	_file.read(header, 3);
	header[3] = '\0';

	if (std::string(header) != "TTR")
	{
		Logger::log_error("Invalid file header: {}", header);
		return false;
	}

	_file.read(reinterpret_cast<char*>(&_offset_to_current_domain), sizeof(_offset_to_current_domain));
	_file.read(reinterpret_cast<char*>(&_offset_to_domains_end), sizeof(_offset_to_domains_end));

	_file.read(reinterpret_cast<char*>(&_offset_to_entities), sizeof(_offset_to_entities));

	_file.seekg(_offset_to_current_domain);
	_file.read(reinterpret_cast<char*>(&_num_of_domains), sizeof(_num_of_domains));

	_file.seekg(_offset_to_entities);
	_file.read(reinterpret_cast<char*>(&_num_of_entities), sizeof(_num_of_entities));

	return true;
}

bool TTRFile::_update_header()
{
	_file.seekp(0);

	_file.write("TTR", 3);

	_file.write(reinterpret_cast<const char*>(&_offset_to_current_domain), sizeof(_offset_to_current_domain));
	_file.write(reinterpret_cast<const char*>(&_offset_to_domains_end), sizeof(_offset_to_domains_end));

	_file.write(reinterpret_cast<const char*>(&_offset_to_entities), sizeof(_offset_to_entities));

	return true;
}